packages/target-web3-v1/src/codegen/types.ts

```diff
@@ -57,18 +66,87 @@ export function codegenOutputType(evmType: EvmOutputType): string {
     case 'dynamic-bytes':
       return 'string'
     case 'array':
-      return `(${codegenOutputType(evmType.itemType)})[]`
+      return codegenArrayOrTupleType(codegenOutputType(options, evmType.itemType), evmType.size)
     case 'boolean':
       return 'boolean'
     case 'string':
       return 'string'
     case 'tuple':
-      return codegenTupleType(evmType, codegenOutputType)
+      if (evmType.structName && options.useStructs) {
+        return evmType.structName.toString() + STRUCT_OUTPUT_POSTFIX
+      }
+      return codegenOutputComplexType(evmType.components, { ...options, useStructs: true })
     case 'unknown':
       return 'any'
   }
 }
 
-export function codegenTupleType(tuple: TupleType, generator: (evmType: EvmType) => string) {
-  return '[' + tuple.components.map((component) => generator(component.type)).join(', ') + ']'
+export function codegenArrayOrTupleType(item: string, length?: number) {
+  if (length !== undefined && length < 6) {
+    return `[${Array(length).fill(item).join(', ')}]`
+  } else {
+    return `${item}[]`
+  }
```

Comment:
Hi @mihai9-lab, we ran across a bug on this line specifically. In the case where item is a string declaring a union type (for example: `string | number | BN`), just appending `[]` to this type results in `string | number | BN[]` instead of the intended type `(string | number | BN)[]` or perhaps `string [] | number[] | BN[]` (I prefer the first solution, as it allows mixing of types in input arrays).

I was able to fix this with the following code:
```
if (item.contains("|")) {
  item = `(${item})`
}
```

However, it may be cleaner to parametrize this function with another boolean and determine whether item is a union upstream of this function, and/or accept length and this proposed boolean as properties of an options object.

Reply:
Seems like the original code here just wrapped all types in parenthesis regardless of whether the type was simple or a union type. I leave the exact form of the solution up to you!